dynparquet/schema.go

```diff
@@ -1351,46 +1352,65 @@ func mergeDynamicRowGroupDynamicColumns(rowGroups []DynamicRowGroup) map[string]
 }
 
 func MergeDynamicColumnSets(sets []map[string][]string) map[string][]string {
-	dynamicColumns := map[string][][]string{}
-	for _, set := range sets {
-		for k, v := range set {
-			_, seen := dynamicColumns[k]
-			if !seen {
-				dynamicColumns[k] = [][]string{}
-			}
-			dynamicColumns[k] = append(dynamicColumns[k], v)
-		}
-	}
+	m := newDynamicColumnSetMerger()
+	defer m.Release()
+	return m.Merge(sets)
+}
 
-	resultDynamicColumns := map[string][]string{}
-	for name, dynCols := range dynamicColumns {
-		resultDynamicColumns[name] = mergeDynamicColumns(dynCols)
-	}
+type dynColSet struct {
+	keys   []string
+	seen   map[string]struct{}
+	values []string
+}
 
-	return resultDynamicColumns
+func newDynamicColumnSetMerger() *dynColSet {
+	return mergeSetPool.Get().(*dynColSet)
 }
 
-// mergeDynamicColumns merges the given concrete dynamic column names into a
-// single superset. It assumes that the given DynamicColumns are all for the
-// same dynamic column name.
-func mergeDynamicColumns(dyn [][]string) []string {
-	return mergeStrings(dyn)
+var mergeSetPool = &sync.Pool{New: func() any {
+	return &dynColSet{
+		seen:   make(map[string]struct{}),
+		keys:   make([]string, 0, 16), // This is arbitrary we anticipate to be lower than values size
+		values: make([]string, 0, 64), // This is arbitrary
+	}
+}}
+
+func (c *dynColSet) Release() {
+	c.keys = c.keys[:0]
+	clear(c.seen)
+	c.values = c.values[:0]
+	mergeSetPool.Put(c)
 }
 
-// mergeStrings merges the given sorted string slices into a single sorted and
-// deduplicated slice of strings.
-func mergeStrings(str [][]string) []string {
-	result := []string{}
-	seen := map[string]struct{}{}
-	for _, s := range str {
-		for _, n := range s {
-			if _, ok := seen[n]; !ok {
-				result = append(result, n)
-				seen[n] = struct{}{}
+func (c *dynColSet) Merge(sets []map[string][]string) (o map[string][]string) {
+	o = make(map[string][]string)
```

Comment:
The nice thing about doing a k-way merge is that you can avoid allocations of "seen" data points. I would implement this using `container/heap` as a crutch. I don't think it's necessary to implement this in this PR, but could you add a TODO?

Reply:
done c3bfff1

I just realised I completely misunderstood @brancz comment.